docs/proposals/20220414-runtime-hooks.md

```diff
@@ -0,0 +1,426 @@
+
+* Operations for installing add-ons during the cluster provisioning workflow
+* Operations for upgrading add-ons during the cluster upgrade workflow
+* Operations for handling add-ons during the cluster deletion workflow
+
+Runtime Hooks enable the Cluster Lifecycle to trigger these processes based on the current state of the cluster, allowing them to start after some state has been reached or waiting for them to complete before moving on with Cluster-wide operations such as Cluster creation or deletion.
+
+Once these hooks are in place, it will be possible to build a comprehensive add-on orchestration solution on top of Cluster API that can leverage external tools such as helm, kapp, ArgoCD, flux or eventually CRS as an alternative.
+
+The hooks defined in this proposal are going to model common events in the Cluster lifecycle and we assume that in the future they may be used for other use cases beyond add-on management. 
+
+### Goals
+
+* Identify a set of Runtime Hooks that enable management of the entire add-on lifecycle 
+* Define the OpenAPI specification of these Runtime Hooks
+* Document when the corresponding Runtime Extensions are called
+* Provide guidelines for developers implementing a corresponding Runtime Extension
+
+### Non-Goals
+
+* Defining all possible Runtime Hooks in Cluster API; this proposal defines only a subset of hooks required for add-on orchestration.
+* Define a full add-on management solution or define detailed steps for solving add-on related problems like CPI migration from in-tree to out-of-tree; this proposal is focused only in providing the foundational capabilities to do so. 
+
+### Future work
+
+* Improve the hooks defined in this proposal to better support use cases beyond add-on management 
+
+
+## Proposal
+
+This proposal adds a set of Runtime Hooks specifically designed for Cluster created from a ClusterClass and thus leveraging the idea of managed topology. 
+
+The main reason for this choice is because a managed topology has a set of capabilities required for lifecycle hooks implementation:
+
+* A managed topology has the overarching view of the entire Cluster (vs other components in CAPI which are limited to controlling single resources or a subset of them).
+* A managed topology already has the capability to control all resources in a Cluster, thus making it possible to orchestrate lifecycle workflows like e.g. upgrades.
+
+Please note that this is especially relevant because it allows treating the cluster lifecycle as a state machine no matter the underlying controller being based on level-triggered reconcile loops.
+
+###  User Stories
+
+These user stories are based on a concrete example of an add-on - a metrics database -  to illustrate the use of each Runtime Hook.
```

Comment:
for being able to reason thoroughly about the right hooks/abstractions can we add a few more concrete examples in addition to the metrics database?

Reply:
@enxebre Do you have any suggestions/pointers for example we can add here?